lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -0,0 +1,225 @@
+module Unison.Sqlite.Transaction
+  ( -- * Transaction management
+    Transaction,
+    runTransaction,
+
+    -- * Executing queries
+
+    -- ** Without results
+
+    -- *** With parameters
+    execute,
+    executeMany,
+
+    -- *** Without parameters
+    execute_,
+
+    -- ** With results
+
+    -- *** With parameters
+    queryList,
+    queryListOne,
+    queryMaybe,
+    queryMaybeOne,
+    queryOne,
+    queryOneOne,
+
+    -- **** With checks
+    queryListCheck,
+    queryListOneCheck,
+    queryMaybeCheck,
+    queryMaybeOneCheck,
+    queryOneCheck,
+    queryOneOneCheck,
+
+    -- *** Without parameters
+    queryList_,
+    queryListOne_,
+    queryMaybe_,
+    queryMaybeOne_,
+    queryOne_,
+    queryOneOne_,
+
+    -- **** With checks
+    queryListCheck_,
+    queryListOneCheck_,
+    queryMaybeCheck_,
+    queryMaybeOneCheck_,
+    queryOneCheck_,
+    queryOneOneCheck_,
+  )
+where
+
+import Control.Monad.Trans.Reader (ReaderT (..))
+import qualified Database.SQLite.Simple as Sqlite
+import qualified Database.SQLite.Simple.FromField as Sqlite
+import Unison.Prelude
+import Unison.Sqlite.Connection (Connection (..))
+import qualified Unison.Sqlite.Connection as Connection
+import Unison.Sqlite.Sql
+
+newtype Transaction a
+  = Transaction (Connection -> IO a)
+  -- Omit MonadIO instance because transactions may be retried
+  -- Omit MonadThrow instance so we always throw SqliteException (via *Check) with lots of context
+  deriving (Applicative, Functor, Monad) via (ReaderT Connection IO)
+
+-- | Run a transaction on the given connection.
+runTransaction :: MonadIO m => Connection -> Transaction a -> m a
+runTransaction conn@(Connection _ _ conn0) (Transaction f) =
+  -- TODO some sensible retry logic
```

Comment:
Retry logic is intentionally omitted from this PR as the logic might be a bit involved, and can come later. It's fine to use a transaction abstraction that doesn't actually retry on `SQLITE_BUSY` in the meantime.